### ISSUE TITLE
IMyPowerProducer ingame script access to MaxPowerOutput from definition

### ISSUE THREAD

**Issue by MMaster** [KeenSoftwareHouse/SpaceEngineers#139]
There is small issue with MaxPowerOutput reported from IMyPowerProducer which reports 0MW when the block is turned off. It makes sense for some cases as when the block is turned off its max output should be 0MW, but most of the time scripts are interested in max power output that it can produce when turned on. Even Details text in battery a reactor shows max value from definition. Also the Solar panel max is always equal current output.

Since I'm not sure what would be acceptable way of exposing this generically for all power producing blocks, I'm opening this issue for suggestions / discussion. 

I think it would be nice if in-game scripts would have access to definitions of blocks so maybe it would be even better to do this even more generically.

**Comment by mexmer** [comment]
I don't think it's good idea for ingame script to have access to block definition per se, but exposing some fields from block definition should be ok.

Still there is problem, that on some blocks, even value from block definition is modified by modifier that is evaluated at runtime.

**Comment by MMaster** [comment]
Why do you think that exposing block definition is not a good idea? Is there something dangerous?
You are right about the modifier issue, but I think details text updates when modifier is active so it should be possible to expose it. Question is what would be the best way so we don't end up with loads of different functions each exposing MaxPowerOutput with little bit different logic behind it.

**Comment by mexmer** [comment]
nah, exposing block defintion is not dangerous (imo), but problem is those modifiers, you need also to expose modifiers, and then people will need to combine those informations to get actual values. imo too much coding and complexity will be rather cumbersome to use.

**Comment by MMaster** [comment]
I understand. So it is better idea to expose functions which take upgrades into account for things that are possibly interesting for ingame scripts. Lets focus on power producers for now. We have IMyPowerProducer which is common for all power producing blocks and contains functions to expose power output values. The only issue is that this does not expose correct max value when the power producing block is turned off + it does not expose correct max value for solar panels. I don't think it is acceptable to modify existing getter and therefore it is necessary to expose new function which will return actual max value instead of 0. Would it make sense to expose new getter in IMyPowerProducer and overriding it in each block type to use max value from definition and take active upgrades into account? If so do you have any suggestion for name of this property?

**Comment by mexmer** [comment]
i will need to look more into Ingame.IPowerProducer, i believe there should not be 2 max values, so either override existing one, or declare existing one obsolete, and create new one, not sure about name tho'

**Comment by MMaster** [comment]
Yes that is what is bothering me too - exposing 2 max values is not right. However I'm afraid to change behavior of existing one in case something depends on it's current behavior. I will try to look go through all references to it and try to find where it is used. If I find out that this change would not break anything then I think I will override existing one and do pull request for it. Otherwise I will probably declare something like MaxRealPowerOutput (I don't like that name tbh please someone suggest better one) and deprecate old one.

**Comment by zrisher** [comment]
MaxPossiblePowerOutput? Potential?

**Comment by mexmer** [comment]
@MMaster if it's properly implemented and defined as Ingame.IMyPowerProducer.MaxPowerOutput, then it should not be problem
